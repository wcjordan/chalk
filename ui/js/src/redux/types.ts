export interface Todo {
  id: number;
  archived: boolean;
  completed: boolean;
  created_at: number;
  description: string;
}

export interface TodoPatch {
  id: number;
  archived?: boolean;
  completed?: boolean;
  created_at?: number;
  description?: string;
}

export interface NewTodo {
  id?: number;
  archived?: boolean;
  completed?: boolean;
  created_at?: number;
  description?: string;
}

export interface ApiState<T> {
  entries: T[];
  loading: boolean;
}

export interface ReduxState {
  todosApi: ApiState<Todo>;
  workspace: WorkspaceState;
}

export interface WorkspaceState {
  editId: number | null;
  labelPickerVisible: boolean;
  labels: string[];
  selectedLabels: { [label: string]: boolean };
}
