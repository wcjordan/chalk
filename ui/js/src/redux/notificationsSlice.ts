import { createSlice } from '@reduxjs/toolkit';

import { NotificationsState } from './types';

const initialState: NotificationsState = {
  notificationQueue: [],
};
export default createSlice({
  name: 'notifications',
  initialState,
  reducers: {
    addNotification: (state, action) => {
      state.notificationQueue.push(action.payload);
    },
    dismissNotification: (state) => {
      state.notificationQueue.shift();
    },
  },
});
