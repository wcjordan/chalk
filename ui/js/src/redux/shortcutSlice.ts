import { createSlice } from '@reduxjs/toolkit';

import { ShortcutState } from './types';

const initialState: ShortcutState = {
  operations: [],
  latestGeneration: 0,
};
export default createSlice({
  name: 'shortcuts',
  initialState,
  reducers: {
    addEditTodoOperation: (state, action) => {
      state.operations.push({
        type: 'EDIT_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    addMoveTodoOperation: (state, action) => {
      state.operations.push({
        type: 'MOVE_TODO',
        payload: action.payload,
        generation: state.latestGeneration,
      });
    },
    incrementGenerations: (state) => {
      state.latestGeneration = state.latestGeneration + 1;
    },
    clearOperationsUpThroughGeneration: (state, action) => {
      state.operations = state.operations.filter(
        (op) => op.generation > action.payload,
      );
    },
  },
});
