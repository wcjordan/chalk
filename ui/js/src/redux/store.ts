import { configureStore } from '@reduxjs/toolkit';
import rootReducer, { listTodos } from './reducers';

const store = configureStore({
  reducer: rootReducer,
});

// TODO dispose
window.setInterval(() => store.dispatch(listTodos()), 1000);

export default store;
