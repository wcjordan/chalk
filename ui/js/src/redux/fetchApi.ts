import Cookies from 'js-cookie';
import { Platform } from 'react-native';

import { getEnvFlags } from '../helpers';
import { RootState } from './store';

const webCsrfToken = Cookies.get('csrftoken') as string;
export function getCsrfToken(getState: () => RootState): string {
  if (Platform.OS === 'web') {
    return webCsrfToken;
  }
  return getState().workspace.csrfToken as string;
}

export async function list<T>(apiUri: string): Promise<Array<T>> {
  const response = await fetch(apiUri, getRequestOpts('GET'));
  return handleResponse<Array<T>>(response);
}

export async function patch<T, Q extends { id: number }>(
  apiUri: string,
  entryPatch: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('PATCH', csrfToken);
  requestOpts.body = JSON.stringify(entryPatch);
  const response = await fetch(`${apiUri}${entryPatch.id}/`, requestOpts);
  return handleResponse<T>(response);
}

export async function create<T, Q>(
  apiUri: string,
  newEntry: Q,
  csrfToken: string,
): Promise<T> {
  return postRequest<T, Q>(apiUri, newEntry, csrfToken);
}

export async function postRequest<T, Q>(
  uri: string,
  data: Q,
  csrfToken: string,
): Promise<T> {
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = JSON.stringify(data);
  const response = await fetch(uri, requestOpts);
  return handleResponse<T>(response);
}

function getRequestOpts(
  method: string,
  csrfToken: string | null = null,
): RequestInit {
  const headers: HeadersInit = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };
  if (csrfToken) {
    headers['X-CSRFToken'] = csrfToken;
  }

  const credentials: RequestCredentials = Platform.select({
    native: 'include',
    default: 'same-origin',
  });

  return {
    credentials: credentials,
    headers: headers,
    method: method,
  };
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (!response.ok) {
    throw new Error(`${response.status} ${response.statusText}`);
  }
  return await response.json();
}

export function getWsRoot(): string {
  if (Platform.OS === 'web') {
    return '';
  }

  const subdomain =
    getEnvFlags().ENVIRONMENT === 'prod' ? 'chalk' : 'chalk-dev';
  return `http://${subdomain}.flipperkid.com/`
}

// Used to exchange login token for session cookie in mobile login flow
export async function completeAuthCallback(token: string) {
  return await fetch(
    `${getWsRoot()}api/todos/auth_callback/?code=${token}`,
    getRequestOpts('GET'),
  );
}

export async function logSessionData(
  dataString: string,
  csrfToken: string,
): Promise<string> {
  
  const requestOpts = getRequestOpts('POST', csrfToken);
  requestOpts.body = dataString;
  const response = await fetch(
    `${getWsRoot()}api/todos/log_session_data/`,
    requestOpts,
  );
  return handleResponse<string>(response);
}
