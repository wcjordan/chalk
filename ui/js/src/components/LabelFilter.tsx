import React, { useCallback } from 'react';
import { StyleSheet, View, ViewStyle } from 'react-native';
import { FilterState } from '../redux/types';
import FilterViewControls from './FilterViewControls';
import LabelChip from './LabelChip';

interface Style {
  labelFilterView: ViewStyle;
  spacer: ViewStyle;
}

const UNLABELED = 'Unlabeled';
const styles = StyleSheet.create<Style>({
  labelFilterView: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    width: '100%',
    backgroundColor: '#d9f0ffff',
  },
  spacer: {
    flexGrow: 1,
  },
});

const LabelFilter: React.FC<Props> = function (props: Props) {
  // TODO (jordan) memoize the conversion
  const {
    labels,
    selectedLabels,
    showCompletedTodos,
    showLabelFilter,
    toggleLabel,
    toggleShowCompletedTodos,
    toggleShowLabelFilter,
  } = props;

  const filterByLabelCb = useCallback(
    (label: string) => {
      toggleLabel(label);
    },
    [toggleLabel],
  );

  const chips = labels.map((label) => (
    <LabelChip
      key={label}
      label={label}
      onPress={filterByLabelCb}
      status={selectedLabels[label]}
    />
  ));
  return (
    <View style={styles.labelFilterView} testID="label-filter">
      {chips}
      <LabelChip
        key={UNLABELED}
        label={UNLABELED}
        onPress={filterByLabelCb}
        status={selectedLabels[UNLABELED]}
      />
      <View style={styles.spacer} />
      <FilterViewControls
        showCompletedTodos={showCompletedTodos}
        showLabelFilter={showLabelFilter}
        toggleShowCompletedTodos={toggleShowCompletedTodos}
        toggleShowLabelFilter={toggleShowLabelFilter}
      />
    </View>
  );
};

type Props = {
  labels: string[];
  selectedLabels: FilterState;
  showCompletedTodos: boolean;
  showLabelFilter: boolean;
  toggleLabel: (label: string) => void;
  toggleShowCompletedTodos: () => void;
  toggleShowLabelFilter: () => void;
};

export default LabelFilter;
