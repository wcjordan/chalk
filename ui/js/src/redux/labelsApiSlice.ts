import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { ApiState, ReduxState, Label } from './types';
import { getWsRoot, list } from './fetchApi';

const API_NAME = 'labelsApi';

// Exported for testing
export function getLabelsApi(): string {
  return `${getWsRoot()}api/todos/labels/`;
}

const initialState: ApiState<Label> = {
  entries: [],
  initialLoad: true,
  loading: false,
};

export const listLabels = createAsyncThunk<
  Label[],
  void,
  { state: ReduxState }
>(`${API_NAME}/list`, async () => list<Label>(getLabelsApi()), {
  condition: (_unused, { getState }) => {
    return !getState().labelsApi.loading;
  },
});

export default createSlice({
  name: API_NAME,
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(listLabels.pending, (state) => {
        state.loading = true;
      })
      .addCase(listLabels.fulfilled, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        state.entries = action.payload;
      })
      .addCase(listLabels.rejected, (state, action) => {
        state.initialLoad = false;
        state.loading = false;
        console.warn(`Loading Labels failed. ${action.error.message}`);
      });
  },
});
