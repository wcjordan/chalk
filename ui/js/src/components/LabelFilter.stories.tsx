import React from 'react';
import { FILTER_STATUS } from '../redux/types';
import LabelFilter from './LabelFilter';

const labels = [
  'low-energy',
  'high-energy',
  'vague',
  'work',
  'home',
  'errand',
  'mobile',
  'desktop',
  'email',
  'urgent',
  '5 minutes',
  '25 minutes',
  '60 minutes',
];
const selectedLabels = {
  '5 minutes': FILTER_STATUS.Active,
  work: FILTER_STATUS.Inverted,
  home: FILTER_STATUS.Active,
  'low-energy': FILTER_STATUS.Active,
  mobile: FILTER_STATUS.Active,
};

export default {
  title: 'Label Filter',
  component: LabelFilter,
};
export const DefaultLabelFilter: React.FC = () => (
  <LabelFilter labels={labels} selectedLabels={selectedLabels} />
);
