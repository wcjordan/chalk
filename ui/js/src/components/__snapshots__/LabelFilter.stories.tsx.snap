// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Filter DefaultLabelFilter smoke-test 1`] = `
<div class="css-g5y9jx r-78d8zt r-18u37iz r-1w6e6rj r-13qz1uu"
     data-testid="label-filter"
>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
       data-testid="chip-active-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip-active"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div class="css-g5y9jx r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
          <div dir="auto"
               role="img"
               tabindex="-1"
               class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
               style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
          >
            󰄬
          </div>
        </div>
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          low-energy
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          high-energy
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          vague
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-11oht4w r-1bq2mok r-105ug2t"
       data-testid="chip-inverted-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip-inverted"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div class="css-g5y9jx r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
          <div dir="auto"
               role="img"
               tabindex="-1"
               class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
               style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
          >
            󰄬
          </div>
        </div>
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          work
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
       data-testid="chip-active-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip-active"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div class="css-g5y9jx r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
          <div dir="auto"
               role="img"
               tabindex="-1"
               class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
               style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
          >
            󰄬
          </div>
        </div>
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          home
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          errand
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
       data-testid="chip-active-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip-active"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div class="css-g5y9jx r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
          <div dir="auto"
               role="img"
               tabindex="-1"
               class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
               style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
          >
            󰄬
          </div>
        </div>
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          mobile
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          desktop
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          email
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          urgent
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-1iizze8 r-1bq2mok r-105ug2t"
       data-testid="chip-active-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip-active"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div class="css-g5y9jx r-1kihuf0 r-tuq35u r-1m04atk r-9aemit">
          <div dir="auto"
               role="img"
               tabindex="-1"
               class="css-146c3p1 r-1niwhzg r-lrvibr r-633pao"
               style="font-size: 18px; color: rgb(29, 25, 43); transform: scaleX(1); line-height: 18px; font-family: material-community; font-weight: normal; font-style: normal;"
          >
            󰄬
          </div>
        </div>
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          5 minutes
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          25 minutes
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          60 minutes
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-1phboty r-18u37iz r-rs99b7 r-78d8zt r-1bq2mok r-105ug2t"
       data-testid="chip-container"
       style="border-color: rgba(0, 0, 0, 0); border-radius: 8px; box-shadow: rgba(0, 0, 0, 0) 0px 0px 0px;"
  >
    <button role="button"
            tabindex="0"
            class="css-g5y9jx r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-13qz1uu"
            data-testid="chip"
            type="button"
            style="border-radius: 8px;"
    >
      <div class="css-g5y9jx r-1awozwy r-18u37iz r-bnwqim r-gy4na3"
           style="padding-right: 0px;"
      >
        <div dir="auto"
             class="css-146c3p1 r-dnmrzs r-1udh08x r-1udbk01 r-3s2u2q r-1iln25a r-fdjqy7 r-ws14 r-9iso6 r-1x2decy r-15zivkp r-1jkjb r-1kb76zh r-14gqq1x r-lrvibr"
             style="direction: ltr; color: rgb(29, 25, 43); letter-spacing: 0.1px; font-weight: 500; line-height: 20px; font-size: 14px;"
        >
          Unlabeled
        </div>
      </div>
    </button>
  </div>
  <div class="css-g5y9jx r-16y2uox">
  </div>
  <div dir="auto"
       class="css-146c3p1 r-1pyaxff"
  >
    <div class="css-g5y9jx r-xoduu5 r-1udh08x r-crgep1 r-105ug2t"
         data-testid="icon-button-container"
         style="background-color: rgba(0, 0, 0, 0); width: 36px; height: 36px; border-width: 0px; border-radius: 18px; border-color: rgb(121, 116, 126);"
    >
      <button aria-disabled="true"
              disabled
              role="button"
              tabindex="-1"
              class="css-g5y9jx r-xoduu5 r-12vffkv r-bnwqim r-ctqt5z r-1udh08x r-113qch9 r-1awozwy r-16y2uox r-1777fci"
              data-testid="icon-button"
              type="button"
      >
        <span role="img"
              tabindex="-1"
              class="css-1jxf684 r-1niwhzg r-lrvibr r-633pao"
              style="font-size: 20px; color: rgb(0, 0, 0); transform: scaleX(1); line-height: 20px; font-family: material-community; font-weight: normal; font-style: normal;"
        >
          󰈲
        </span>
      </button>
    </div>
    <div class="css-g5y9jx r-xoduu5 r-1udh08x r-crgep1 r-105ug2t"
         data-testid="show-labels-container"
         style="background-color: rgba(0, 0, 0, 0); width: 36px; height: 36px; border-width: 0px; border-radius: 18px; border-color: rgb(121, 116, 126);"
    >
      <button role="button"
              tabindex="0"
              class="css-g5y9jx r-xoduu5 r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-1awozwy r-16y2uox r-1777fci"
              data-testid="show-labels"
              type="button"
      >
        <span role="img"
              tabindex="-1"
              class="css-1jxf684 r-1niwhzg r-lrvibr r-633pao"
              style="font-size: 20px; color: rgb(0, 0, 0); transform: scaleX(1); line-height: 20px; font-family: material-community; font-weight: normal; font-style: normal;"
        >
          󱋷
        </span>
      </button>
    </div>
    <div class="css-g5y9jx r-xoduu5 r-1udh08x r-crgep1 r-105ug2t"
         data-testid="show-completed-todos-container"
         style="background-color: rgba(0, 0, 0, 0); width: 36px; height: 36px; border-width: 0px; border-radius: 18px; border-color: rgb(121, 116, 126);"
    >
      <button role="button"
              tabindex="0"
              class="css-g5y9jx r-xoduu5 r-1otgn73 r-1loqt21 r-bnwqim r-ctqt5z r-1udh08x r-1awozwy r-16y2uox r-1777fci"
              data-testid="show-completed-todos"
              type="button"
      >
        <span role="img"
              tabindex="-1"
              class="css-1jxf684 r-1niwhzg r-lrvibr r-633pao"
              style="font-size: 20px; color: rgb(0, 0, 0); transform: scaleX(1); line-height: 20px; font-family: material-community; font-weight: normal; font-style: normal;"
        >
          󰄱
        </span>
      </button>
    </div>
  </div>
</div>
`;
